keps/draft-20180426-namespace-template.md

```diff
@@ -0,0 +1,513 @@
+
+1. Put a label that is excluded in the `labelSelector` of the `NamespaceTemplate`
+2. Annotate the namespace with `policy/namespace-template-opt-out=true`
+
+Note: the objects previously populated by the controller will be left as-is in the namespace when it's opted out.
+
+Normally, there's no _exclusive_ condition in `labelSelector` of the `NamespaceTemplate`, so option 2 is recommended.
+
+Some ephemeral namespaces created by privileged controllers may need to be opt-out
+as they are not accessibly by users and will be managed completely by the controllers.
+As these controllers don't have knowledge about `NamespaceTemplate`, 
+or it's almost impossible to customize the way how they create namespaces,
+an automated opt-out mechanism is required.
+The mechanism is based on the service account running the controllers.
+With a `MutatingAdmissionWebhook`, 
+a namespace created by these controllers will be automatically annotated for opt-out.
+This mechanism is out of MVP scope.
+
+#### Self-serviced namespace creation
+
+Self-serviced namespace creation allows certain users/groups to create namespaces
+by themselves using Kubernetes API (or kubectl) directly.
+As only cluster admin has the privilege to create namespaces by default, an extra
+`ClusterRole` is created to allow _create_ verb of namespaces:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: namespace-creator
+rules:
+  - apiGroups: [""]
+    resources: ["namespaces"]
+    verbs: ["create"]
+```
+
+After the namespace is created, 
+the creator should be granted further permissions to manage the namespace.
+The following objects are recommended to be included in the `templates` of `NamespaceTemplate`:
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: namespace-owner
+rules:
+  - apiGroups: [""]
+    resources: ["namespaces"]
+    resourceNames: ["$(NAMESPACE)"]
+    verbs: ["get", "list", "watch", "update", "patch", "delete"]
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: admin
+roleRef:
+  kind: ClusterRole
+  name: admin
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: User
+  name: "$(CREATOR)"
+  apiGroup: rbac.authorization.k8s.io
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: namespace-owner
+roleRef:
+  kind: Role
+  name: namespace-owner
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: User
+  name: "$(CREATOR)"
+  apiGroup: rbac.authorization.k8s.io
+```
+
+There is one issue that when the controller gets informed to create objects inside a new
+namespace from `NamespaceTemplate`, it has no idea who created the namespace.
+To be able to substitute `$(CREATOR)`, a `MutatingAdmissionWebhook` is involved in namespace
+creation request. 
+It adds an annotation `policy/namespace-creator=name` to the namespace being created.
+This annotation is immutable from further update/patch requests.
```

Comment:
This is a pretty big deal, it means that any naive code that changes the annotation on namespaces will break when this is introduced.  I'm not sure we should be doing a one off "this annotation is special" approach without a bit more discussion on next steps.